### PR TITLE
lineage 1/n: add support for metadata.

### PR DESCRIPTION
**Pull request by eric-anderson**
* Add new MetadataDocument subclass that stores metadata and raises errors if tried to be used as a document.
* Add BaseMap class that processes documents, passes through metadata, and automatically generate lineage links.

### CODE REVIEW

File: lib/sycamore/sycamore/data/document.py [hunk 1]
```diff
@@ -155,18 +171,104 @@ def deserialize(raw: bytes) -> "Document":
...
+
+    @property
+    def metadata(self) -> dict[str, Any]:
+        """Internal metadata babout A collection of system or customer defined metadata, for instance,
+        a PDF document might have title and author metadata."""
+        return self.data.get("metadata", {})
+
+    @metadata.setter
+    def metadata(self, metadata: dict[str, Any]):
+        """Set all the proprites for this document."""
```
Comment (bsowell) [inline lib/sycamore/sycamore/data/document.py:263]: minor: proprites -> properties

File: lib/sycamore/sycamore/transforms/base.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    return decorator
+
+
+class BaseMap(UnaryNode):
+    """
+    BaseMap abstracts away MetadataDocuments from all other transforms.
+    """
+
+    def __init__(
```
Comment (bsowell) [inline lib/sycamore/sycamore/transforms/base.py:30]: Once this is used everywhere, does this replace all of the map functions in https://github.com/aryn-ai/sycamore/blob/main/lib/sycamore/sycamore/transforms/map.py?

File: lib/sycamore/sycamore/transforms/base.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        args = _noneOr(self._args, tuple())
+        kwargs = _noneOr(self._kwargs, {})
+
+        @rename(name)
+        def ray_callable(ray_input: dict[str, np.ndarray]) -> dict[str, list]:
+            return BaseMap._process_ray(ray_input, name, lambda d: f(d, *args, **kwargs))
+
+        return ray_callable
+
+    def _map_class(self):
```
Comment (bsowell) [inline lib/sycamore/sycamore/transforms/base.py:76]: One thing I'm trying to understand is whether this has any impact on when Ray will create an actor or a task? At the moment I believe we use an actor when you try to map a class (which I think is from [here](https://github.com/aryn-ai/sycamore/blob/13322f80ec3610063b1073b14b1d006f44d0e8c1/lib/sycamore/sycamore/transforms/map.py#L37)). I don't know that we have to do it exactly the way we do it now, but would be good to understand if there are changes.

**Review by bsowell — APPROVED** [review]
Mostly minor typos. One more substantive question that might be more relevant for the next PR. I don't need to see this one again unless you think there is something we should discuss.

File: lib/sycamore/sycamore/transforms/base.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+def rename(new_function_name: str):
+    def decorator(f):
+        f.__name__ = new_function_name
+        return f
+
+    return decorator
+
+
+class BaseMap(UnaryNode):
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/transforms/base.py:25]: I went with BaseMapTransform.

File: lib/sycamore/sycamore/data/document.py [hunk 5]
```diff
@@ -155,18 +171,104 @@ def deserialize(raw: bytes) -> "Document":
...
+        raise ValueError("MetadataDocument does not have properties")
+
+    @properties.setter
+    def properties(self, properties: dict[str, Any]):
+        raise ValueError("MetadataDocument does not have properties")
+
+    @property
+    def metadata(self) -> dict[str, Any]:
+        """Internal metadata babout A collection of system or customer defined metadata, for instance,
+        a PDF document might have title and author metadata."""
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/data/document.py:258]: Fixed. Looks like I copied and pasted and just stopped filling it in.

File: lib/sycamore/sycamore/data/document.py [hunk 6]
```diff
@@ -16,6 +17,9 @@ def __init__(self, document=None, /, **kwargs):
             from pickle import loads
 
             document = loads(document)
+            if "metadata" in document:
+                raise ValueError("metadata must be unserialized with Document.unserialize not Document.__init__")
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/data/document.py:21]: fixed

File: lib/sycamore/sycamore/data/document.py (hunk 1 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/data/document.py:263]: fixed.

File: lib/sycamore/sycamore/transforms/base.py (hunk 2 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/transforms/base.py:30]: As discussed, probably not. The map functions in map.py become thin wrappers around this class.

File: lib/sycamore/sycamore/transforms/base.py (hunk 3 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/transforms/base.py:76]: It is not intended to have an impact. In class mode it makes a single instance and calls it repeatedly (as verified by test). I think that's what the old code also does.

**Pull request merged**